File: Algorithms/0109_Convert_Sorted_List_to_Binary_Search_Tree/Python/Convert_Sorted_List_to_Binary_Search_Tree_Solution_1.py

# Space: O(n)
# Time: O(n)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def sortedListToBST(self, head):
        if head is None: return None
        if head.next is None: return TreeNode(head.val)
        cache = []
        pointer = head
        while pointer:
            cache.append(pointer.val)
            pointer = pointer.next

        def build_tree(alist):
            if len(alist) == 0: return None
            if len(alist) == 1: return TreeNode(alist[0])
            mid = len(alist) // 2
            head = TreeNode(alist[mid])
            left = alist[:mid]
            right = alist[mid + 1:]
            head.left = build_tree(left)
            head.right = build_tree(right)
            return head

        return build_tree(cache)

File: Algorithms/0109_Convert_Sorted_List_to_Binary_Search_Tree/Java/ConvertSortedListToBinarySearchTreeSolution1.java

// Space: O(n)
// Time: O(n)

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        if (head == null) return null;
        if (head.next == null) return new TreeNode(head.val);
        List<Integer> cache = new ArrayList<>();
        while (head != null) {
            cache.add(head.val);
            head = head.next;
        }
        return buildTree(cache);
    }

    private TreeNode buildTree(List<Integer> ints) {
        if (ints.size() == 0) return null;
        if (ints.size() == 1) return new TreeNode(ints.get(0));
        int mid = ints.size() / 2;
        TreeNode head = new TreeNode(ints.get(mid));
        head.left = buildTree(ints.subList(0, mid));
        head.right = buildTree(ints.subList(mid + 1, ints.size()));
        return head;
    }
}




